File: solns/clojure/4.clj
(defn is-palindrome?
  [n]
  (let [s (str n)]
    (= s (apply str (reverse s)))))

(def result
  (reduce max
    (for [i (range 999 1 -1)
          j (range i   1 -1)
          :let [x (* i j)]
          :when (is-palindrome? x)]
      x)))

(println result)
(assert (= result 906609))

File: solns/python/4.py
from functools import reduce


def is_palindrome(s):
    return s == s[::-1]


def gen():
    for i in range(1000, 1, -1):
        for j in range(i + 1, 1, -1):
            s = i * j
            if is_palindrome(str(s)):
                yield s


result = reduce(max, gen())

print(result)
assert result == 906609
